infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -62,159 +78,160 @@ public class SequencerServerCache implements AutoCloseable {
+            evictSmallestTxVersion();
         }
-        final int numPqEntries = pqEntries;
-        MicroMeterUtils.measure(numPqEntries, "sequencer.cache.evictions");
-        log.info("Invalidated entries {} addresses {}", pqEntries, entries);
     }
 
     /**
-     * The cache size as the number of entries
-     *
-     * @return cache size
+     * Evict all records up to a trim mark (not included).
+     * @param trimMark The trim mark.
      */
-    public int size() {
-        return conflictKeys.size();
+    public void evictUpTo(long trimMark) {
+        final SortedMap<Long, Set<ConflictTxStream>> entriesToEvict = versionMap.headMap(trimMark);
+        final long numTxVersionsToEvict = entriesToEvict.size();
+        long numConflictKeysEvicted = 0;
+
+        if (numTxVersionsToEvict > 0) {
+            log.debug("evictUpTo: trim mark {}", trimMark);
+            for (int l = 0; l < numTxVersionsToEvict; l++) {
+                numConflictKeysEvicted += evictSmallestTxVersion();
+            }
+        }
+
+        log.info("evictUpTo: entries={} addresses={}", numConflictKeysEvicted, numTxVersionsToEvict);
+        MicroMeterUtils.measure(numConflictKeysEvicted, "sequencer.cache.evictions");
     }
 
     /**
-     * The memory space used by the entries and also the space used by
-     * priority queue and hashmap to store the pointers
-     * @return the memory space used in bytes:
+     * The cache size as the number of entries.
+     * @return The current cache size.
      */
-    public long byteSize() {
-        log.debug("the cache has {} entries,  the object size used {}, calculated by beepSize {}",
-                size(), CONFLICTTXSTREAM_OBJ_SIZE,
-                cacheEntries.isEmpty() ? 0 : Memory.sizeOf.deepSizeOf(cacheEntries.peek()));
-        return size() * (ENTRY_OVERHEAD + CONFLICTTXSTREAM_OBJ_SIZE);
+    public int size() {
+        return conflictKeyMap.size();
     }
 
-    /*
-     * Put a value in the cache
-     *
-     * @param conflictStream conflict stream
+    /**
+     * Put the provided hashed conflict keys into the cache,
+     * and evict older entries as necessary.
+     * @param conflictKeys The conflict keys to add into the cache.
+     * @param txVersion    The timestamp associated with these conflict keys.
      */
-    public boolean put(ConflictTxStream conflictStream) {
-
-        Long val = conflictKeys.getOrDefault(conflictStream, Address.NON_ADDRESS);
-        if (val > conflictStream.txVersion) {
-            log.error("For key {} the new entry address {} is smaller than the entry " +
-                            "address {} in cache. There is a sequencer regression.",
-                    conflictStream, conflictStream.txVersion, conflictKeys.get(conflictStream));
-            return false;
-        }
+    public void put(@NonNull Set<ConflictTxStream> conflictKeys, long txVersion) {
+        // The provided timestamp should not already be present in the cache.
```

Comment:
Can you also put a check that makes sure that the conflict keys can actually fit in the cache window. In other words, the cache must hold all of the transactions conflict keys. For example, imagine the cache size = 10 and a transaction tries to commit 20 conflict keys.

Reply:
Added this precondition check and a test for it.